add create index API as an async job

pdowler opened the pull request:
this proposed solution is for a parameter-style UWS job that fits the current DAL landscape

pdowler committed:
add create index API as an async job
this proposed solution is for a parameter-style UWS job that fits the
current DAL landscape

mbtaylor commented on openapi/vosi-table-ops.yaml:
```diff
@@ -0,0 +1,50 @@
+            $ref: ../vosi/vosi-std-responses.yaml#/permission-denied
+        404:
+            $ref: ./vosi-std-responses.yaml#/not-found
+        405:
+            $ref: ./vosi-std-responses.yaml#/not-implemented
```
405 is not likely to be an appropriate response here - it's only correct if this resource exists but does not support any POST requests (and would have to be accompanied by an `Allow` header giving a list of HTTP methods that are supported for that resource) - see RFC9110 sec 15.5.6.

pdowler replied:
Do you think that a 404 when probing for an optional endpoint is sufficient for the client? 

This is a UWS job list endpoint so I agree that optional means it is either there or not, but technically the possibility of a 405 is really up to UWS. That's kind of the awkward thing about UWS and OpenAPI: the job description (params) have to be here but the response can come from the UWS component.

brianmajor replied:
I guess the job list (GET) would also have to return a 404.  Seems natural to me because a 1.1 TAP implementation would return a 404.

pdowler replied:
405 removed, 404 specified

pdowler committed:
remove 405 response and fix references to other vosi components

pdowler committed:
Merge branch 'main' into table-ops

pdowler committed:
Merge branch 'main' of https://github.com/ivoa-std/VOSI into table-ops

pdowler committed:
restructure openapi from main

pdowler committed:
add openapi-linter and rename openapi-example
add global tags to openapi example

pdowler committed:
update table-ops openapi spec to include INDEX_TYPE
add table-ops job and phase using UWS pattern

pdowler committed:
add copy of uws-responses.yaml to repo

pdowler committed:
add reusable UWS job list params to openapi/uws

pdowler committed:
refactor more common uws components

pdowler commented:
Additional comment about the UWS content included here:

As with TAP-async itself, UWS is a pattern that services implement and I have come to the conclusion that it is not possible to write components that can simply be pulled in and used as-is. 

already known: the `vosi-tables-ops.yaml` is a UWS endpoint and the GET operation can mostly be defined in terms of reusable parts, but the POST operation defines the job-description-language (params in this case) for this endpoint. In the TAP standard there will be an equivalent yaml that defines TAP-async (different params). 

But: also note that each operation has an `operationId` and that has to be unique when an arbitrary set of components are pulled together. Further, you will now see included here `vosi-tables-ops-job.yaml` which has to differ from a generic uws-job.yaml by having a unique `operationId`. So, in this case (and in the TAP-async case) one must materialise the uws resources and assign unique operationId values so they are properly nested/connected and do not collide.

The reusable parts of UWS I have identified and included in openapi/uws/

pdowler committed:
combine unique into INDEX_TYPE param
specify multi-values explicitly with type: array
specify serialisation with explode
fix UWS WAIT param example

pdowler committed:
tweak to table-ops INDEX parameter

pdowler commented:
additional detail on the use of `style: form` (default), `explode: true|false`, and `type: array|string`.

`type: array` allows for multiple values
`explode: true` means those multiple values are in multiple param=value pairs
`explode=false` means those values are in one param with comma-separated values

Here in the "create index job" we can specify something that works like this (sure, doesn't necessarily make sense and might not be allowed):
```
TABLE=myschema.mytable
INDEX=ra,dec
INDEX_TYPE=long-lat
INDEX_TYPE=unique
```
 but we can specify one comma-separated param and one normal multi param.

That seems fine, but DALI UPLOAD param actually does both:
- you can have multiple UPLOAD params (so type: array, explode: true)
- each value is a comma-separated name,location pair
So  we have prior art for comma-separated values that are definitely *not* what OpenAPI explode=false is describing. I do need to fix the DALI upload yaml (type, explode) and there I will stick with the regex pattern to specify the content.

So here, I have reverted INDEX to a simpler `type: string` (single value) and use the description to specify "comma-separated list of column names". I could add a pattern (regex) which constrains values if that's worthwhile, but seems minimal value (for code gen).

pdowler committed:
openapi lint improvements
